test/fail_compilation/test18385.d

```diff
@@ -0,0 +1,31 @@
+/*
+REQUIRED_ARGS: -de
+TEST_OUTPUT:
+---
+fail_compilation/test18385.d(13): Deprecation: function `test18385.foo` function `test18385.foo(double)` cannot be overloaded with another `extern(C)` function at fail_compilation/test18385.d(12)
+fail_compilation/test18385.d(18): Deprecation: function `test18385.S.foo` function `test18385.S.foo(double)` cannot be overloaded with another `extern(C)` function at fail_compilation/test18385.d(17)
+---
+*/
+
+extern (C):
+
+void foo(int) { }
+void foo(double) { }
+
+struct S
+{
+    static void foo(int) {}
+    static void foo(double) {}
+}
+
+void foo2(int) { }
+extern(D) void foo2(double) { } // OK as it has a different mangling
+
+void foo3(int) { }
+void foo3(double); // duplicate declarations are allowed
```

Comment:
(so you probably wanted to write `void foo3(int);` ?)

Reply:
I guess the rational is that this is `extern(C)` which would mangle to the same function.

Reply:
argh... disregard my comments on this piece. Perhaps explicitly add `extern(C)` in front of each function here? Makes it easier to read the testcases.

Reply:
I'm not convinced this avoid-explicit-casts use-case is worth the ugly inconsistent AST, and having to handle incompatible declarations, incl. casts, in the backend.

Reply:
I'm not sure I follow. This PR just tries to enable a deprecation of being more restrictive than dmd is currently. It was originally intended for 2.079, but got disabled/reverted due to people already complaining about it.
I agree that the avoid-explicit-casts use-case is a pretty bad one and I would love to deprecate more ;-)

Reply:
Ah sorry I meant the test; I'd rather have it check that a compatible declaration (`void foo3(int);`) and later definition works, instead of testing that incompatible declarations are allowed.